### ISSUE TITLE
Figure out how to handle users with empty names

### ISSUE THREAD

**Issue by JosteinLindhom** [quickfeed/quickfeed#896]

**Comment by JosteinLindhom** [comment]
Some thoughts:

Currently users with empty names may appear in the list of users. This happens when a user does not complete the signup process, i.e. fill out their profile with name, email, and student ID.

To avoid this we could, upon user creation, set their name and email to the ones retrieved from GitHub. This would still require users to set their student ID. 

However, in many cases a user will not have set a name on GitHub. Additionally, the user would have to go through extra steps to update their name on QuickFeed if it does not match their full name.

Currently a user is restricted from navigating away from the profile form if their user information has not been filled in. This can be circumvented if someone truly wants to, enabling a user to enroll in courses. Additionally, nothing stops a user from sending a `CreateEnrollment` request using their cookie without having filled in their user information.

To avoid the scenarios in the previous paragraph, we could disallow creation of enrollments where a database user does not have (name, email, student_id) set.

If we allow users to enroll in courses only if their user is valid, and disallow nullifying any of the user (name, email, student_id) fields, we could safely purge incomplete users from our database. This could be done alongside #308

**Comment by meling** [comment]
The above stuff sounds good to me. Although, not sure it is worth the extra effort to prevent enrollments... but maybe it is easy (didn't check).

But what if we add a check along these lines in auth.go:
```go
func checkExternalUser(externalUser *externalUser) error {
	if externalUser.Login == "" {
		return errors.New("missing login")
	}
	if externalUser.Name == "" {
		return errors.New("missing name")
	}
	if externalUser.Email == "" {
		return errors.New("missing email")
	}
	return nil
}
```
We could check that the name has at least two components (first and last name), and that the email is a proper email, and if this check fails, we **refuse to create an account** on quickfeed. Hopefully the frontend would somehow explain what we didn't like about their GitHub account profile.

We could even require that they use a `@uis.no` email, but that is perhaps a bit too much.

**Comment by meling** [comment]
Yes, I agree with your concerns about privacy. Looks like the best approach would be to implement Canvas integration.

**Comment by JoachimTislov** [comment]
> To avoid the scenarios in the previous paragraph, we could disallow creation of enrollments where a database user does not have (name, email, student_id) set.

Sounds good, along with this:

> 
```go 
func checkExternalUser(externalUser *externalUser) error {
        if externalUser.Login == "" {
	        return errors.New("missing login")
	}
	if externalUser.Name == "" {
		return errors.New("missing name")
	}
	if externalUser.Email == "" {
		return errors.New("missing email")
	}
	return nil
}
```
> We could check that the name has at least two components (first and last name), and that the email is a proper email, and if this check fails, we refuse to create an account on quickfeed. Hopefully the frontend would somehow explain what we didn't like about their GitHub account profile.

Yes, an even better solution would be to prevent the user from creating an account all together.


@JosteinLindhom How many accounts has had empty names or private email addresses?

Cause I agree we everything mentioned above, and I think its weird how Github even allows it. For privacy reasons, sure, but empty names??
